pkg/appgw/requestroutingrules.go

```diff
@@ -59,131 +59,11 @@ func (c *appGwConfigBuilder) RequestRoutingRules(cbCtx *ConfigBuilderContext) er
-			}
-		}
-
-		for ruleIdx := range ingress.Spec.Rules {
-			rule := &ingress.Spec.Rules[ruleIdx]
-			if rule.HTTP == nil {
-				// skip no http rule
-				continue
-			}
-
-			listenerHTTPID := generateListenerID(rule, n.HTTP, nil, usePrivateIPForIngress)
-			_, httpAvailable := httpListenersMap[listenerHTTPID]
-
-			listenerHTTPSID := generateListenerID(rule, n.HTTPS, nil, usePrivateIPForIngress)
-			_, httpsAvailable := httpListenersMap[listenerHTTPSID]
-
-			if httpAvailable {
-				if wildcardRule != nil && len(rule.Host) != 0 {
-					// only add wildcard rules when host is specified
-					urlPathMaps[listenerHTTPID] = c.pathMaps(ingress, cbCtx, wildcardRule,
-						listenerHTTPID, urlPathMaps[listenerHTTPID],
-						defaultAddressPoolID, defaultHTTPSettingsID)
-				}
-
-				// need to eliminate non-unique paths
-				urlPathMaps[listenerHTTPID] = c.pathMaps(ingress, cbCtx, rule,
-					listenerHTTPID, urlPathMaps[listenerHTTPID],
-					defaultAddressPoolID, defaultHTTPSettingsID)
-
-				// If ingress is annotated with "ssl-redirect" and we have TLS - setup redirection configuration.
-				if sslRedirect, _ := annotations.IsSslRedirect(ingress); sslRedirect && httpsAvailable {
-					c.modifyPathRulesForRedirection(cbCtx, urlPathMaps[listenerHTTPID], listenerHTTPSID)
-				}
-			}
-
-			if httpsAvailable {
-				if wildcardRule != nil && len(rule.Host) != 0 {
-					// only add wildcard rules when host is specified
-					urlPathMaps[listenerHTTPSID] = c.pathMaps(ingress, cbCtx, wildcardRule,
-						listenerHTTPSID, urlPathMaps[listenerHTTPSID],
-						defaultAddressPoolID, defaultHTTPSettingsID)
-				}
-
-				// need to eliminate non-unique paths
-				urlPathMaps[listenerHTTPSID] = c.pathMaps(ingress, cbCtx, rule,
-					listenerHTTPSID, urlPathMaps[listenerHTTPSID],
-					defaultAddressPoolID, defaultHTTPSettingsID)
-			}
-		}
-	}
-
-	if len(urlPathMaps) == 0 {
-		defaultAddressPoolID := c.appGwIdentifier.addressPoolID(defaultBackendAddressPoolName)
-		defaultHTTPSettingsID := c.appGwIdentifier.httpSettingsID(defaultBackendHTTPSettingsName)
-		listenerID := defaultFrontendListenerIdentifier()
-		urlPathMaps[listenerID] = &n.ApplicationGatewayURLPathMap{
-			Etag: to.StringPtr("*"),
-			Name: to.StringPtr(generateURLPathMapName(listenerID)),
-			ApplicationGatewayURLPathMapPropertiesFormat: &n.ApplicationGatewayURLPathMapPropertiesFormat{
-				DefaultBackendAddressPool:  &n.SubResource{ID: &defaultAddressPoolID},
-				DefaultBackendHTTPSettings: &n.SubResource{ID: &defaultHTTPSettingsID},
-				PathRules:                  &[]n.ApplicationGatewayPathRule{},
-			},
-		}
-	}
-
-	return urlPathMaps
-}
-
 func (c *appGwConfigBuilder) getRules(cbCtx *ConfigBuilderContext) ([]n.ApplicationGatewayRequestRoutingRule, []n.ApplicationGatewayURLPathMap) {
-	httpListenersMap := c.groupListenersByListenerIdentifier(c.getListeners(cbCtx))
+	httpListenersMap := c.groupListenersByListenerIdentifier(c.appGw.HTTPListeners)
```

Comment:
This is making the assumption that we have `c.appGw.HTTPListeners` already - but I don't think we have guarantees for that here.  If we keep this line - we would have an implicitly required order of execution, which is not a good thing.

I propose we keep `c.getListeners(cbCtx)` here and if performance is a concern - we memoize it.

Reply:
Here's a tweak to alleviate the side-effects concerts (duplicate log lines) - https://github.com/Azure/application-gateway-kubernetes-ingress/pull/445/files